feature-assets/src/main/java/io/novafoundation/nova/feature_assets/presentation/novacard/overview/NovaCardViewModel.kt

```diff
@@ -0,0 +1,112 @@
+import io.novafoundation.nova.feature_assets.presentation.novacard.topup.TopUpCardPayload
+import io.novafoundation.nova.feature_wallet_api.presentation.model.toAssetPayload
+import io.novafoundation.nova.runtime.ext.ChainGeneses
+import io.novafoundation.nova.runtime.ext.utilityAsset
+import io.novafoundation.nova.runtime.multiNetwork.ChainRegistry
+import io.novafoundation.nova.runtime.multiNetwork.chain.model.Chain
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import java.math.BigDecimal
+import kotlinx.coroutines.flow.map
+
+class NovaCardViewModel(
+    private val chainRegistry: ChainRegistry,
+    private val accountInteractor: AccountInteractor,
+    private val assetsRouter: AssetsRouter,
+    private val novaCardInteractor: NovaCardInteractor,
+    private val cardCreationInterceptorFactory: CardCreationInterceptorFactory,
+    private val topUpRequestInterceptorFactory: TopUpRequestInterceptorFactory,
+    private val novaCardWebViewControllerFactory: NovaCardWebViewControllerFactory
+) : BaseViewModel(), CardCreationInterceptor.Callback, TopUpRequestInterceptor.Callback {
+
+    private val openedOrderIds = mutableSetOf<String>()
+
+    private val metaAccount = flowOf { accountInteractor.selectedMetaAccount() }
+
+    private val setupCardConfig = metaAccount.map { metaAccount ->
+        val topUpChain = getTopUpChain()
+        CardSetupConfig(
+            refundAddress = metaAccount.requireAddressIn(topUpChain),
+            spendToken = topUpChain.utilityAsset
+        )
+    }
+
+    val novaCardWebViewControllerFlow = setupCardConfig.map { setupConfig ->
+        novaCardWebViewControllerFactory.create(
+            interceptors = listOf(
+                cardCreationInterceptorFactory.create(this),
+                topUpRequestInterceptorFactory.create(this)
+            ),
+            setupConfig = setupConfig,
+            scope = viewModelScope
+        )
+    }
+
+    init {
+        checkCardCreationState()
+    }
+
+    override fun onTopUpStart(orderId: String, amount: BigDecimal, address: String) {
+        if (orderId in openedOrderIds) return // To not handle same order twice
+        openedOrderIds.add(orderId)
+
+        launch {
+            val payload = TopUpCardPayload(
+                amount = amount,
+                address = address,
+                asset = setupCardConfig.first().spendToken.toAssetPayload()
+            )
+
+            assetsRouter.openTopUpCard(payload)
+        }
+    }
+
+    override fun onTopUpCompleted(orderId: String) {
+        launch {
+            openedOrderIds.remove(orderId)
+
+            novaCardInteractor.setTopUpFinishedEvent()
+            setCardStateCreated()
+        }
+    }
+
+    override fun onCardCreated() {
+        setCardStateCreated()
+    }
+
+    private suspend fun getTopUpChain(): Chain {
+        return chainRegistry.getChain(ChainGeneses.POLKADOT)
+    }
+
+    private fun checkCardCreationState() {
```

Comment:
I think the name of the function should indicate that it might open a bottom sheet
Something like `ensureCardCreationIsBlocking`?